sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -779,6 +786,43 @@ def fit(self, X, y, sample_weight=None):
         del self._in_fit  # hard delete so we're sure it can't be used anymore
         return self
 
+    def _ordinal_encode_df(self, X, return_categorical_features=False):
+        if (
+            self.categorical_features is None
+            and hasattr(X, "columns")
+            and hasattr(X, "assign")
+        ):
+            # pandas dataframe: find explicitly typed categorical columns,
+            # extract their inner ordinal encoding and record the indices of
+            # the categorical columns before calling _validate_data and
+            # _check_categorical_features later.
+            if return_categorical_features:
+                categorical_features = np.zeros(X.shape[1], dtype=bool)
+            for col_idx, col in enumerate(X.columns):
+                if X[col].dtype.name == "category":
+                    # As of now, missing values are always encoded as np.nan in
+                    # this estimator, therefore we need to use a floating point
+                    # dtype to encode the categories.
+                    # XXX: using integers would be more efficient but would
+                    # make the code more complex.
+                    codes = X[col].values.codes.astype(np.float64)
```

Comment:
Here are my thoughts on using panda's `codes`:

1. **Mismatched cateogires**: I think it is important to check that the categorical dtype are the same during `fit` and `transform`, otherwise the `codes` encoding may represent different categories. For example, if the fit time categorical dtype is `['dog', 'cat', 'turtle']` and predict time is `['cat', 'turtle', 'dog']`, then codes `[0, 1, 2]` will mean different things.

2. **Unknown Categories**: In the following, "bear" is unknown:

```python
X_train["pet"].cat.categories
# ["cat", "dog", "snake", ...]

X_test["pet"].cat.categories
# ["cat", "bear", "dog", "snake", ...]
```

During train time `"dog"` is 1, but during predict time `"bear"` is 1. We can technically do some preprocessing here to map the unknown categories to missing. The simplest solution is to check the categorical dtypes are the same during `fit` and `predict`.

3. **Infrequent Categories**: In the following, "bear" is infrequent and happened to only be in the train set:

```python
X_train["pet"].cat.categories
# ["cat", "bear", "dog", "snake", ...]

X_test["pet"].cat.categories
# ["cat", "dog", "snake", ...]
```

In this case, the "bear" is in the training set, but not in the test set. We can handle this by some preprocessing to correctly remap the codes during `test` to the ones seen in `train`. Again, the simplest solution is to check the categorical dtype are the same during `fit` and `predict`

**TLDR**: I think checking that the categories match during `fit` and `predict` is the simplest way to work around all three situations.

Reply:
> TLDR: I think checking that the categories match during fit and predict is the simplest way to work around all three situations.

I agree. I will update this PR later this week. Or feel free to push to this PR directly if you prefer as I don't plan to work on this today nor tomorrow.

Reply:
I am on vacation this week, so I am trying my best not to write code. I will be available next week.

Reply:
I added the consistency check for pandas categorical features here: [`00f6a18` (#24907)](https://github.com/scikit-learn/scikit-learn/pull/24907/commits/00f6a18b54b8d962c81def87608ad1f35850e36f)